rust/core-lib/src/linewrap.rs

```diff
@@ -405,63 +407,146 @@ impl<'a> Iterator for VisualLines<'a> {
 
 /// A cursor over both hard and soft breaks. Currently this is either/or,
 /// but eventually soft will be soft only, and this will interleave them.
+///
+/// # Invariants:
+///
+/// `self.offset` is always a valid break in one of the cursors, unless
+/// at 0 or EOF.
+///
+/// `self.offset == self.text.pos().min(self.soft.pos())`.
 struct MergedBreaks<'a> {
     text: Cursor<'a, RopeInfo>,
     soft: Cursor<'a, BreaksInfo>,
     offset: usize,
+    /// Starting from zero, how many calls to `next` to get to `self.offset`?
+    cur_line: usize,
+    total_lines: usize,
 }
 
 impl<'a> Iterator for MergedBreaks<'a> {
     type Item = usize;
 
     fn next(&mut self) -> Option<usize> {
-        // TODO: remove when soft and hard are separate
+        if self.text.pos() == self.offset && self.offset != self.text.total_len() {
```

Comment:
Looks like `self.offset != self.text.total_len()` is used elsewhere too, maybe turn that into an `at_end()`?

Reply:
yea, good call.